File: LeetCode/Java/42-stack-接雨水.java

/*
 * @lc app=leetcode.cn id=42 lang=java
 *
 * [42] 接雨水
 */
import java.util.*;

// @lc code=start
class Solution {
    public int trap(int[] height) {
        int res = 0;
        int current = 0;
        var stack = new Stack<Integer>();
        while (current < height.length) {
            while (!stack.empty() && height[current] > height[stack.peek()]) {
                int top = stack.pop();
                if (stack.empty())
                    break;
                int distance = current - stack.peek() - 1;
                int boundedHeight = Math.min(height[current], height[stack.peek()]) - height[top];
                res += distance * boundedHeight;
            }
            stack.push(current++);
        }
        return res;
    }
}
// @lc code=end

File: LeetCode/C/42-stack-接雨水.cpp
/*
 * @lc app=leetcode.cn id=42 lang=cpp
 *
 * [42] 接雨水
 */
// #include <bits/stdc++.h>
// using namespace std;

// @lc code=start
class Solution {
  public:
    int trap(vector<int> &height) {
        int ans = 0, current = 0;
        stack<int> st;
        while (current < height.size()) {
            while (!st.empty() && height[current] > height[st.top()]) {
                int top = st.top();
                st.pop();
                if (st.empty())
                    break;
                int distance = current - st.top() - 1;
                int bounded_height =
                    min(height[current], height[st.top()]) - height[top];
                ans += distance * bounded_height;
            }
            st.push(current++);
        }
        return ans;
    }
};
// @lc code=end
